File: examples/typescript/yds/function/consumer.ts
import { Handler } from '@yandex-cloud/function-types';

/**
 * Consumer function triggered by YDS (Yandex Data Streams).
 *
 * Processes batches of messages from the YDS topic.
 * Messages are automatically decoded and parsed.
 */

interface ProcessedMessage {
    user_id: string;
    action: string;
    status: string;
}

interface UserAction {
    action: string;
    message: string;
    timestamp: string;
}

export const handler: Handler.DataStreams = async (event, context) => {
    try {
        const messages = event.messages || [];
        console.log(`Received ${messages.length} messages from YDS`);

        // Debug: Log the full event structure
        console.log(`Full event: ${JSON.stringify(event, null, 2)}`);

        if (messages.length === 0) {
            return {
                statusCode: 200,
                body: JSON.stringify({ status: 'success', processed: 0 })
            };
        }

        // Process messages
        const processedMessages: ProcessedMessage[] = [];
        const userActions: Record<string, UserAction[]> = {};

        for (const msg of messages) {
            try {
                // Debug: Log the message structure
                console.log(`Message structure: ${JSON.stringify(msg, null, 2)}`);

                // Extract fields
                const userId = msg.user_id;
                const action = msg.action;
                const message = msg.message;
                const timestamp = msg.timestamp;

                console.log(`Processing message: user=${userId}, action=${action}, message=${message}`);

                // Group actions by user
                if (!userActions[userId]) {
                    userActions[userId] = [];
                }
                userActions[userId].push({
                    action,
                    message,
                    timestamp
                });

                // Process based on action type
                processAction(userId, action, message, timestamp);

                processedMessages.push({
                    user_id: userId,
                    action,
                    status: 'processed'
                });
            } catch (error) {
                console.error('Error processing individual message:', error);
                console.error('Error stack:', error instanceof Error ? error.stack : String(error));
                continue;
            }
        }

        // Log batch summary
        console.log(`Batch processing complete. Processed ${processedMessages.length} messages`);
        console.log(`User summary: ${JSON.stringify(userActions)}`);

        // Print summary by user
        for (const [userId, actions] of Object.entries(userActions)) {
            const actionList = actions.map(a => a.action).join(', ');
            console.log(`User ${userId}: ${actions.length} actions - [${actionList}]`);
        }

        return {
            statusCode: 200,
            body: JSON.stringify({
                status: 'success',
                processed: processedMessages.length,
                users_affected: Object.keys(userActions).length
            })
        };

    } catch (error) {
        console.error('Error in consumer handler:', error);
        return {
            statusCode: 500,
            body: JSON.stringify({
                error: error instanceof Error ? error.message : String(error)
            })
        };
    }
};

/**
 * Process different action types with custom business logic.
 *
 * @param userId - User identifier
 * @param action - Action type (login, logout, purchase, view, etc.)
 * @param message - Message content
 * @param timestamp - Message timestamp
 */
function processAction(userId: string, action: string, message: string, timestamp: string): void {
    switch (action) {
        case 'login':
            console.log(`User ${userId} logged in at ${timestamp}`);
            // Add custom logic for login events
            // e.g., update user session, send analytics, etc.
            break;

        case 'logout':
            console.log(`User ${userId} logged out at ${timestamp}`);
            // Add custom logic for logout events
            break;

        case 'purchase':
            console.log(`User ${userId} made a purchase: ${message}`);
            // Add custom logic for purchase events
            // e.g., update inventory, send confirmation email, etc.
            break;

        case 'view':
            console.log(`User ${userId} viewed: ${message}`);
            // Add custom logic for view events
            // e.g., track page views, update recommendations, etc.
            break;

        default:
            console.log(`User ${userId} performed action '${action}': ${message}`);
            // Handle other action types
            break;
    }
}

File: examples/python/yds/function/consumer.py
import json
import logging
import base64
from collections import defaultdict

# Configure logging
logger = logging.getLogger()
logger.setLevel(logging.INFO)


def consumer_handler(event, context):
    """
    Consumer function triggered by YDS (Yandex Data Streams).

    Processes batches of messages from the YDS topic.
    Messages are automatically decoded and parsed.

    Event structure:
    {
        "messages": [
            {
                "event_metadata": {
                    "event_id": "string",
                    "event_type": "string",
                    "created_at": "timestamp"
                },
                "details": {
                    "topic": "string",
                    "partition": "string",
                    "database": "string",
                    "message": {
                        "data": "base64_encoded_string",
                        "seq_no": "string",
                        "created_at": "timestamp",
                        "message_group_id": "string"
                    }
                }
            }
        ]
    }
    """
    try:
        messages = event.get('messages', [])
        logger.info(f"Received {len(messages)} messages from YDS")

        if not messages:
            return {
                'statusCode': 200,
                'body': json.dumps({'status': 'success', 'processed': 0})
            }

        # Process messages
        processed_messages = []
        user_actions = defaultdict(list)

        for msg in messages:
            try:
                # Extract message data
                details = msg.get('details', {})
                message_data = details.get('message', {})
                data_encoded = message_data.get('data', '')

                # Decode base64 data
                if data_encoded:
                    data_decoded = base64.b64decode(data_encoded).decode('utf-8')
                    data_json = json.loads(data_decoded)

                    # Extract fields
                    user_id = data_json.get('user_id')
                    action = data_json.get('action')
                    message = data_json.get('message')
                    timestamp = data_json.get('timestamp')

                    logger.info(f"Processing message: user={user_id}, action={action}, message={message}")

                    # Group actions by user
                    user_actions[user_id].append({
                        'action': action,
                        'message': message,
                        'timestamp': timestamp
                    })

                    # Process based on action type
                    process_action(user_id, action, message, timestamp)

                    processed_messages.append({
                        'user_id': user_id,
                        'action': action,
                        'status': 'processed'
                    })

            except Exception as e:
                logger.error(f"Error processing individual message: {e}")
                continue

        # Log batch summary
        logger.info(f"Batch processing complete. Processed {len(processed_messages)} messages")
        logger.info(f"User summary: {dict(user_actions)}")

        # Print summary by user
        for user_id, actions in user_actions.items():
            logger.info(f"User {user_id}: {len(actions)} actions - {[a['action'] for a in actions]}")

        return {
            'statusCode': 200,
            'body': json.dumps({
                'status': 'success',
                'processed': len(processed_messages),
                'users_affected': len(user_actions)
            })
        }

    except Exception as e:
        logger.error(f"Error in consumer handler: {e}")
        return {
            'statusCode': 500,
            'body': json.dumps({'error': str(e)})
        }


def process_action(user_id, action, message, timestamp):
    """
    Process different action types with custom business logic.

    Args:
        user_id: User identifier
        action: Action type (login, logout, purchase, view, etc.)
        message: Message content
        timestamp: Message timestamp
    """
    if action == 'login':
        logger.info(f"User {user_id} logged in at {timestamp}")
        # Add custom logic for login events
        # e.g., update user session, send analytics, etc.

    elif action == 'logout':
        logger.info(f"User {user_id} logged out at {timestamp}")
        # Add custom logic for logout events

    elif action == 'purchase':
        logger.info(f"User {user_id} made a purchase: {message}")
        # Add custom logic for purchase events
        # e.g., update inventory, send confirmation email, etc.

    elif action == 'view':
        logger.info(f"User {user_id} viewed: {message}")
        # Add custom logic for view events
        # e.g., track page views, update recommendations, etc.

    else:
        logger.info(f"User {user_id} performed action '{action}': {message}")
        # Handle other action types
